src/states/joining_peer.rs

```diff
@@ -0,0 +1,300 @@
+                    self, msg
+                );
+                self.msg_backlog.push(msg);
+                Ok(Transition::Stay)
+            }
+        }
+    }
+
+    fn handle_node_approval(&mut self, gen_pfx_info: GenesisPfxInfo) -> Transition {
+        info!(
+            "{} - This node has been approved to join the network!",
+            self
+        );
+        Transition::IntoAdult { gen_pfx_info }
+    }
+
+    #[cfg(feature = "mock_base")]
+    pub fn get_timed_out_tokens(&mut self) -> Vec<u64> {
+        self.timer.get_timed_out_tokens()
+    }
+}
+
+impl Base for JoiningPeer {
+    fn network_service(&self) -> &NetworkService {
+        &self.network_service
+    }
+
+    fn network_service_mut(&mut self) -> &mut NetworkService {
+        &mut self.network_service
+    }
+
+    fn full_id(&self) -> &FullId {
+        &self.full_id
+    }
+
+    fn in_authority(&self, dst: &Authority<XorName>) -> bool {
+        dst.is_single() && dst.name() == *self.full_id.public_id().name()
+    }
+
+    fn min_section_size(&self) -> usize {
+        self.min_section_size
+    }
+
+    fn peer_map(&self) -> &PeerMap {
+        &self.peer_map
+    }
+
+    fn peer_map_mut(&mut self) -> &mut PeerMap {
+        &mut self.peer_map
+    }
+
+    fn handle_send_message(
+        &mut self,
+        _: Authority<XorName>,
+        _: Authority<XorName>,
+        _: Vec<u8>,
+    ) -> Result<(), InterfaceError> {
+        warn!("{} - Cannot handle SendMessage - not joined.", self);
+        // TODO: return Err here eventually. Returning Ok for now to
+        // preserve the pre-refactor behaviour.
+        Ok(())
+    }
+
+    fn handle_timeout(&mut self, token: u64, _: &mut dyn EventBox) -> Transition {
+        if self.join_token == token {
+            debug!("{} - Timeout when trying to join a section.", self);
+
+            let connected_addrs = self
+                .peer_map
+                .connected_ids()
+                .filter_map(|pub_id| self.peer_map.get_connection_info(pub_id))
+                .map(|conn_info| conn_info.peer_addr())
+                .collect::<Vec<_>>();
+            for peer_addr in connected_addrs {
+                self.network_service
+                    .service_mut()
+                    .disconnect_from(peer_addr);
+            }
+
+            return Transition::Rebootstrap;
+        }
+
+        Transition::Stay
+    }
+
+    fn handle_connection_failure(
+        &mut self,
+        peer_addr: SocketAddr,
+        _: &mut dyn EventBox,
+    ) -> Transition {
+        let _ = self.peer_map_mut().disconnect(peer_addr);
+
+        Transition::Stay
+    }
+
+    fn handle_direct_message(
+        &mut self,
+        msg: DirectMessage,
+        _pub_id: PublicId,
+        _outbox: &mut dyn EventBox,
+    ) -> Result<Transition, RoutingError> {
+        debug!("{} Unhandled direct message: {:?}", self, msg);
+
+        Ok(Transition::Stay)
+    }
+
+    fn handle_hop_message(
```

Comment:
Joining peer may never receive such messages, although IIRC if it did like Adult it should forward to the section Elders to let them decide what to do with it.

Reply:
It needs to receive exactly one type of such messages - the message that is currently called `NodeApproval`, which triggers its transition into an Adult.